lte/gateway/c/session_manager/SessionState.cpp

```diff
@@ -1175,73 +1175,60 @@ bool SessionState::reset_reporting_charging_credit(
   if (it == credit_map_.end()) {
     // new credit
-    return init_charging_credit(update, update_criteria);
+    return init_charging_credit(update, session_uc);
   }
   auto& grant    = it->second;
-  auto credit_uc = get_credit_uc(CreditKey(update), update_criteria);
+  auto credit_uc = get_credit_uc(CreditKey(update), session_uc);
   if (!update.success()) {
     // update unsuccessful, reset credit and return
-    MLOG(MDEBUG) << session_id_ << " Received an unsuccessful update for RG "
-                 << update.charging_key();
+    MLOG(MERROR) << "Credit update failed RG:" << key << " for " << session_id_;
     grant->credit.mark_failure(update.result_code(), credit_uc);
     if (grant->should_deactivate_service()) {
       grant->set_service_state(SERVICE_NEEDS_DEACTIVATION, *credit_uc);
     }
     return false;
   }
-  MLOG(MINFO) << session_id_ << " Received a charging credit for RG: "
-              << update.charging_key();
+  MLOG(MINFO) << "Received a credit RG:" << key << " for " << session_id_;
   grant->receive_charging_grant(update.credit(), credit_uc);
 
   if (grant->reauth_state == REAUTH_PROCESSING) {
     grant->set_reauth_state(REAUTH_NOT_NEEDED, *credit_uc);
   }
   if (!grant->credit.is_quota_exhausted(1) &&
       grant->service_state != SERVICE_ENABLED) {
-    // if quota no longer exhausted, reenable services as needed
-    MLOG(MINFO) << "Quota available. Activating service";
+    // if quota no longer exhausted, re-enable services as needed
+    MLOG(MINFO) << "New quota now available. Activating service RG:" << key
+                << " for " << session_id_;
     grant->set_service_state(SERVICE_NEEDS_ACTIVATION, *credit_uc);
   }
-  return contains_credit(update.credit().granted_units()) ||
-         is_infinite_credit(update);
+  return true;
 }
 
 bool SessionState::init_charging_credit(
     const CreditUpdateResponse& update,
-    SessionStateUpdateCriteria& update_criteria) {
+    SessionStateUpdateCriteria& session_uc) {
+  const uint32_t key = update.charging_key();
   if (!update.success()) {
     // init failed, don't track key
-    MLOG(MERROR) << "Credit init failed for imsi " << imsi_
-                 << " and charging key " << update.charging_key();
+    MLOG(MERROR) << "Credit init failed RG:" << key << " for " << session_id_;
     return false;
   }
-  MLOG(MINFO) << session_id_ << " Initialized a charging credit for RG: "
-              << update.charging_key();
-
-  auto charging_grant    = std::make_unique<ChargingGrant>();
-  charging_grant->credit = SessionCredit(SERVICE_ENABLED, update.limit_type());
-
-  charging_grant->receive_charging_grant(update.credit());
-  update_criteria.charging_credit_to_install[CreditKey(update)] =
-      charging_grant->marshal();
-  credit_map_[CreditKey(update)] = std::move(charging_grant);
-  return contains_credit(update.credit().granted_units()) ||
-         is_infinite_credit(update);
-}
-
-bool SessionState::contains_credit(const GrantedUnits& gsu) {
```

Comment:
I think it is better not to remove this function and just remove the check for volumes (or if you want to keep the code as it is, just change `>` for `=>`

A non valid granted component doesn't mean 0

Reply:
hmmmmmm yeah that is a good point. I suppose we could still receive a MSCC result code = 2001 but there is no grant AVP? 
Ok I will modify the PR to keep one function to verify the credit.

Reply:
@uri200 Now the updated PR will
1. Reject credits with empty GSU (all credits are is_valid=false). This means the credit does not get added to the credit map and no rules are installed.
2. Valid credits with all 0 values, with at least one is_valid=true, will get added to the credit pool. the rules will also be installed.

I've just moved the `contains_credit` and `infinite_credit` functions to `ChargingGrant` class as it is only relevant for Gy.